dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java

```diff
@@ -81,15 +117,69 @@ private Authentication getAuthentication(HttpServletRequest request) {
+                        res.sendError(HttpServletResponse.SC_BAD_REQUEST, "The login as feature is not allowed" +
+                            " due to the current configuration");
+                        return Pair.of(null, true);
+                    }
+                }
 
                 //Return the Spring authentication object
-                return new DSpaceAuthentication(eperson.getEmail(), authorities);
+                return Pair.of(new DSpaceAuthentication(eperson.getEmail(), authorities), false);
             } else {
-                return null;
+                return Pair.of(null, false);
+            }
+        } else {
+            if (request.getHeader(ON_BEHALF_OF_REQUEST_PARAM) != null) {
+                res.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Only admins are allowed to use the" +
+                    " login as feature");
+                return Pair.of(null, true);
             }
         }
 
-        return null;
+        return Pair.of(null, false);
+    }
+
+    private Pair<Authentication, Boolean> getOnBehalfOfAuthentication(Context context, String onBehalfOfParameterValue,
+                                                       HttpServletResponse res) throws IOException {
+
+        try {
+            if (!authorizeService.isAdmin(context)) {
+                res.sendError(HttpServletResponse.SC_FORBIDDEN, "Only admins are allowed to use the" +
+                    " login as feature");
+                return Pair.of(null, true);
+            }
+            UUID epersonUuid = UUIDUtils.fromString(onBehalfOfParameterValue);
+            if (epersonUuid == null) {
+                res.sendError(HttpServletResponse.SC_BAD_REQUEST, "The given UUID in the X-On-Behalf-Of header " +
+                    "was not a proper UUID");
+                return Pair.of(null, true);
+            }
+            EPerson onBehalfOfEPerson = ePersonService.find(context, epersonUuid);
+            if (onBehalfOfEPerson == null) {
+                res.sendError(HttpServletResponse.SC_BAD_REQUEST, "The given UUID in the X-On-Behalf-Of header " +
+                    "was not a proper EPerson UUID");
+                return Pair.of(null, true);
+            }
+            if (!authorizeService.isAdmin(context, onBehalfOfEPerson)) {
+                requestService.setCurrentUserId(epersonUuid);
+                context.switchContextUser(onBehalfOfEPerson);
+                return Pair.of(new DSpaceAuthentication(onBehalfOfEPerson.getEmail(),
+                                authenticationProvider.getGrantedAuthorities(context, onBehalfOfEPerson)), false);
+            } else {
+                res.sendError(HttpServletResponse.SC_BAD_REQUEST, "You're unable to use the login as feature to log " +
+                    "in as another admin");
+                return Pair.of(null, true);
+            }
+
+        } catch (SQLException e) {
+            log.error(e.getMessage(), e);
+            return Pair.of(null, false);
```

Comment:
This also seems a bit odd to me.  Why would you not return an error when an error occurred?  This seems to be essentially a no-op if a SQLException happens (or am I missing something)?